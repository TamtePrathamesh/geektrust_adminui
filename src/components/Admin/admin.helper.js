/* eslint-disable no-undef */
import { REACT_APP_ADMIN_URL } from '../../services/constants';

export const getAdminList = async () => {
  try {
    let res = await fetch(REACT_APP_ADMIN_URL);
    res = await res.json();
    return res;
  } catch (error) {
    console.error(error);
  }
};
