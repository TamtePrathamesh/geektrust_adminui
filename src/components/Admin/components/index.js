export { default as AdminList } from './AdminList';
export { default as SearchAdmin } from './SearchAdmin';
export { default as DeleteSelected } from './DeleteSelected';
