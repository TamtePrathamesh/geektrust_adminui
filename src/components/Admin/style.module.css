@import '../../style/colors.css';

.adminContainer {
  padding: 20px;
  background-color: var(--white);
}

.bottomSide {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-evenly;
}

@media screen and (max-width: 700px) {
  .adminContainer {
    padding: 10px;
    background-color: var(--white);
  }

  .bottomSide {
    gap: 15px;
    width: 100%;
    display: flex;
    align-items: flex-start;
    flex-direction: column;
    padding-top: 15px;
  }
  .paginated {
    width: 95%;
    padding: 2%;
    overflow-y: scroll;
    border: 1px solid var(--grey-one);
  }
}
