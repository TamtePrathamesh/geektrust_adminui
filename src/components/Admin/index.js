/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useCallback, useReducer } from 'react';
import { AdminList, SearchAdmin, DeleteSelected } from './components';
import { WithBoundary, VerticalGap, Pagination } from '../common';
import { getAdminList } from './admin.helper';

import Style from './style.module.css';

const initialState = {
  offset: 0,
  perPage: 10,
  currentPage: 1,
  pageCount: 0,
  admins: [],
  filter: '',
};

const reducer = (state, { type, value }) => {
  switch (type) {
    case 'state':
      return { ...state, ...value };
    default:
      return initialState;
  }
};

const Admin = () => {
  const [{ pageCount, currentPage, perPage, offset, admins, filter }, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    if (admins.length === 0) {
      const getAdmins = async () => {
        try {
          const res = await getAdminList();
          let updated = res.map((i) => {
            return { ...i, selection: false };
          });
          dispatch({
            type: 'state',
            value: {
              pageCount: Math.ceil(updated.length / perPage),
              admins: updated,
              paginatedData: res.slice(offset, offset + perPage),
            },
          });
        } catch (error) {
          console.error(error);
        }
      };
      getAdmins();
    }
  }, []);

  const getPaginatedData = useCallback(() => {
    return admins.slice(offset, offset + perPage);
  }, [admins, offset, perPage]);

  const onFilter = useCallback((e) => {
    let value = e.target.value;
    if (value) {
      dispatch({ type: 'state', value: { filter: value } });
    } else {
      dispatch({ type: 'state', value: { filter: '' } });
    }
  }, []);

  const onDelete = useCallback(
    (id) => {
      if (id) {
        let filtered = admins.filter(({ id: userId }) => userId !== id);
        dispatch({
          type: 'state',
          value: {
            pageCount: Math.ceil(filtered.length / perPage),
            admins: filtered,
          },
        });
      }
    },
    [admins, perPage]
  );

  const onEdit = useCallback(
    (id, admin) => {
      if (id && admin) {
        let updatedAdmins = admins.map((ad) => (ad.id === id ? Object.assign({}, ad, admin) : ad));
        dispatch({ type: 'state', value: { admins: updatedAdmins } });
      }
    },
    [admins]
  );

  const onSelectOrDeselectAll = useCallback(
    (selection) => {
      let update = admins.map((admin, index) => {
        if (index >= offset && index < offset + perPage) {
          return { ...admin, selection: selection.target.checked };
        }
        return admin;
      });
      dispatch({ type: 'state', value: { admins: [...update] } });
    },
    [admins, offset, perPage]
  );

  const onSelectOrDeselect = useCallback(
    (id, selection) => {
      if (id) {
        let updated = admins.map((ad) => (ad.id === id ? { ...ad, selection: selection.target.checked } : ad));
        dispatch({ type: 'state', value: { admins: updated } });
      }
    },
    [admins]
  );

  const deleteSelected = useCallback(() => {
    let updated = admins.filter(({ selection }) => !selection);
    dispatch({
      type: 'state',
      value: {
        pageCount: Math.ceil(updated.length / perPage),
        admins: updated,
      },
    });
  }, [admins, perPage]);

  const handlePageClick = useCallback(
    (selected, pageCounts) => {
      const selectedPage = selected;
      const calculatedOffset = selectedPage * perPage - perPage;
      dispatch({
        type: 'state',
        value: {
          currentPage: pageCounts < currentPage ? pageCount : selectedPage,
          offset: calculatedOffset,
          paginatedData: admins.slice(calculatedOffset, calculatedOffset + perPage),
        },
      });
    },
    [admins, perPage]
  );

  return (
    <div className={Style.adminContainer}>
      <SearchAdmin onFilter={onFilter} />
      <VerticalGap height="15" />
      {admins && (
        <AdminList
          adminList={getPaginatedData()}
          filter={filter}
          onDelete={onDelete}
          onEdit={onEdit}
          onSelectOrDeselectAll={onSelectOrDeselectAll}
          onSelectOrDeselect={onSelectOrDeselect}
        />
      )}
      <div className={Style.bottomSide}>
        <DeleteSelected deleteSelected={deleteSelected} />
        <div className={Style.paginated}>
          <Pagination pageCount={pageCount} currentPage={currentPage} onPageChange={handlePageClick} />
        </div>
      </div>
    </div>
  );
};

export default WithBoundary(Admin);
