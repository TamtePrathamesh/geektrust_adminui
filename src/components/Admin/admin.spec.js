import React from 'react';
import { render, waitFor, screen, act } from '@testing-library/react';
import user from '@testing-library/user-event';
import * as helper from './admin.helper';
import Admin from '.';

jest.mock('./admin.helper', () => ({
  getAdminList: jest.fn(() => Promise.resolve([])),
}));

describe('Admin', () => {
  it('should match snapshot', async () => {
    helper.getAdminList.mockImplementation(() =>
      Promise.resolve([
        { id: '1', name: 'xyz', email: 'an@g.c', role: 'member', selection: false },
        { id: '2', name: 'pqr', email: 'an@g.c', role: 'member', selection: false },
      ])
    );
    const { container } = render(<Admin />);
    await waitFor(() => expect(helper.getAdminList).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });

  it('should display the data', async () => {
    helper.getAdminList.mockClear();
    helper.getAdminList.mockImplementation(() =>
      Promise.resolve([
        { id: '1', name: 'xyz', email: 'an@g.c', role: 'member', selection: false },
        { id: '2', name: 'pqr', email: 'an@g.c', role: 'member', selection: false },
      ])
    );
    const { container } = render(<Admin />);
    await waitFor(() => expect(helper.getAdminList).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
  });

  it('should delete the Row from data when row delete button clicked', async () => {
    helper.getAdminList.mockClear();
    helper.getAdminList.mockImplementation(() =>
      Promise.resolve([
        { id: '1', name: 'xyz', email: 'an@g.c', role: 'member', selection: false },
        { id: '2', name: 'pqr', email: 'an@g.c', role: 'member', selection: false },
      ])
    );
    render(<Admin />);
    await waitFor(() => expect(helper.getAdminList).toHaveBeenCalledTimes(1));
    expect(screen.queryAllByText('xyz').length).toBe(1);
    user.click(screen.queryAllByTestId('deleteRow')[0]);
    expect(screen.queryAllByText('xyz').length).toBe(0);
  });

  it('should selectAll Rows when select all clicked', async () => {
    helper.getAdminList.mockClear();
    helper.getAdminList.mockImplementation(() =>
      Promise.resolve([
        { id: '1', name: 'xyz', email: 'an@g.c', role: 'member', selection: false },
        { id: '2', name: 'pqr', email: 'an@g.c', role: 'member', selection: false },
      ])
    );
    const { rerender, container } = render(<Admin />);
    await waitFor(() => expect(helper.getAdminList).toHaveBeenCalledTimes(1));
    expect(container.querySelectorAll('tr[class=selectedRow]').length).toBe(0);
    user.click(screen.queryAllByTestId('selectAll')[0]);
    rerender(<Admin />);
    expect(container.querySelectorAll('tr[class=selectedRow]').length).toBe(2);
  });

  it('should delete all selected row when deleteSelected button clicked', async () => {
    helper.getAdminList.mockClear();
    helper.getAdminList.mockImplementation(() =>
      Promise.resolve([
        { id: '1', name: 'xyz', email: 'an@g.c', role: 'member', selection: false },
        { id: '2', name: 'pqr', email: 'an@g.c', role: 'member', selection: false },
      ])
    );
    const { rerender, container } = render(<Admin />);
    await waitFor(() => expect(helper.getAdminList).toHaveBeenCalledTimes(1));
    expect(container.querySelectorAll('tr[class=selectedRow]').length).toBe(0);
    act(() => {
      user.click(screen.queryAllByTestId('selectAll')[0]);
    });
    rerender(<Admin />);
    expect(container.querySelectorAll('tr[class=selectedRow]').length).toBe(2);
    act(() => {
      user.click(screen.queryByTestId('deleteSelected'));
    });
    await waitFor(() => expect(container.querySelectorAll('tr[class=selectedRow]').length).toBe(0));
  });

  it('should save edited row data', async () => {
    helper.getAdminList.mockClear();
    helper.getAdminList.mockImplementation(() =>
      Promise.resolve([
        { id: '1', name: 'xyz', email: 'an@g.c', role: 'member', selection: false },
        { id: '2', name: 'pqr', email: 'an@g.c', role: 'member', selection: false },
      ])
    );
    const { container } = render(<Admin />);
    await waitFor(() => expect(helper.getAdminList).toHaveBeenCalledTimes(1));
    expect(screen.queryAllByText('member').length).toBe(2);
    await user.click(screen.queryAllByTestId('editRow')[0]);
    await user.type(container.querySelector('input[value=member]'), 'admin');
    await user.click(screen.queryByTestId('saveRow'));
    expect(screen.queryAllByText('member').length).toBe(1);
  });
});
