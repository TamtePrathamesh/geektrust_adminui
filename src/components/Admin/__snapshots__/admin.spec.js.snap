// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin should display the data 1`] = `
<div>
  <div
    class="adminContainer"
  >
    <input
      class="search"
      data-testid="search"
      placeholder="Search by name, email or role"
      type="text"
    />
    <div
      class="gap"
      style="height: 15px;"
    />
    <div
      class="table-responsive-sm"
    >
      <table
        class="table"
      >
        <thead>
          <tr>
            <th>
              <input
                checked=""
                data-testid="selectAll"
                type="checkbox"
              />
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Name
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Email
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Role
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
            data-testid="tr"
          >
            <td>
              <input
                data-testid="selectRow"
                type="checkbox"
              />
            </td>
            <td
              class="columnName"
            >
              xyz
            </td>
            <td
              class="columnName"
            >
              an@g.c
            </td>
            <td
              class="columnName"
            >
              member
            </td>
            <td>
              <div
                class="actionItems"
              >
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="editRow"
                  src=""
                />
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="deleteRow"
                  src=""
                />
              </div>
            </td>
          </tr>
          <tr
            class=""
            data-testid="tr"
          >
            <td>
              <input
                data-testid="selectRow"
                type="checkbox"
              />
            </td>
            <td
              class="columnName"
            >
              pqr
            </td>
            <td
              class="columnName"
            >
              an@g.c
            </td>
            <td
              class="columnName"
            >
              member
            </td>
            <td>
              <div
                class="actionItems"
              >
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="editRow"
                  src=""
                />
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="deleteRow"
                  src=""
                />
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="bottomSide"
    >
      <div
        class="container"
      >
        <button
          class="deleteSelected"
          data-testid="deleteSelected"
        >
          Delete Selected
        </button>
      </div>
      <div
        class="paginated"
      >
        <div
          class="paginationContainer"
        >
          <div
            class="pagination"
            data-testid="leading"
          >
            <div
              class="actionBtn"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
            <div
              class="actionBtn disableClick"
              data-testid="previous"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
            <div
              class="contentOverflow"
            >
              <div>
                <button
                  class="stepper active"
                  id="stepper"
                >
                  1
                </button>
              </div>
            </div>
            <div
              class="actionBtn disableClick"
              data-testid="next"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
            <div
              class="actionBtn"
              data-testid="trailing"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Admin should match snapshot 1`] = `
<div>
  <div
    class="adminContainer"
  >
    <input
      class="search"
      data-testid="search"
      placeholder="Search by name, email or role"
      type="text"
    />
    <div
      class="gap"
      style="height: 15px;"
    />
    <div
      class="table-responsive-sm"
    >
      <table
        class="table"
      >
        <thead>
          <tr>
            <th>
              <input
                checked=""
                data-testid="selectAll"
                type="checkbox"
              />
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Name
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Email
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Role
            </th>
            <th
              class="columnName"
              scope="col"
            >
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class=""
            data-testid="tr"
          >
            <td>
              <input
                data-testid="selectRow"
                type="checkbox"
              />
            </td>
            <td
              class="columnName"
            >
              xyz
            </td>
            <td
              class="columnName"
            >
              an@g.c
            </td>
            <td
              class="columnName"
            >
              member
            </td>
            <td>
              <div
                class="actionItems"
              >
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="editRow"
                  src=""
                />
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="deleteRow"
                  src=""
                />
              </div>
            </td>
          </tr>
          <tr
            class=""
            data-testid="tr"
          >
            <td>
              <input
                data-testid="selectRow"
                type="checkbox"
              />
            </td>
            <td
              class="columnName"
            >
              pqr
            </td>
            <td
              class="columnName"
            >
              an@g.c
            </td>
            <td
              class="columnName"
            >
              member
            </td>
            <td>
              <div
                class="actionItems"
              >
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="editRow"
                  src=""
                />
                <img
                  alt="cant display"
                  class="actionIcons"
                  data-testid="deleteRow"
                  src=""
                />
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="bottomSide"
    >
      <div
        class="container"
      >
        <button
          class="deleteSelected"
          data-testid="deleteSelected"
        >
          Delete Selected
        </button>
      </div>
      <div
        class="paginated"
      >
        <div
          class="paginationContainer"
        >
          <div
            class="pagination"
            data-testid="leading"
          >
            <div
              class="actionBtn"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
            <div
              class="actionBtn disableClick"
              data-testid="previous"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
            <div
              class="contentOverflow"
            >
              <div>
                <button
                  class="stepper active"
                  id="stepper"
                >
                  1
                </button>
              </div>
            </div>
            <div
              class="actionBtn disableClick"
              data-testid="next"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
            <div
              class="actionBtn"
              data-testid="trailing"
            >
              <img
                alt="cant display"
                height="10px"
                src=""
                width="10px"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
